// @hint/hint-typescript-config/dist/tests/import-helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const mock = require("mock-require");
const sinon = require("sinon");
const utils = require("@hint/utils");
const utils_tests_helpers_1 = require("@hint/utils-tests-helpers");
const utils_types_1 = require("@hint/utils-types");
const hintPath = utils_tests_helpers_1.getHintPath(__filename, true);
const tests = [
    {
        after: (t) => {
            t.context.sandbox.restore();
        },
        before: (t) => {
            const sandbox = sinon.createSandbox();
            sandbox.stub(utils, 'loadPackage').returns({ exists: true });
            t.context.sandbox = sandbox;
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage: utils.loadPackage });
        },
        name: 'Configuration with "compilerOptions.importHelpers = true" should pass',
        path: path.join(__dirname, 'fixtures', 'import-helpers', 'import')
    },
    {
        after: (t) => {
            t.context.sandbox.restore();
        },
        before: (t) => {
            const sandbox = sinon.createSandbox();
            sandbox.stub(utils, 'loadPackage').throws(new Error('Not found'));
            t.context.sandbox = sandbox;
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage: utils.loadPackage });
        },
        name: 'Configuration with "compilerOptions.importHelpers = true" but tslibs is not installed should fail',
        path: path.join(__dirname, 'fixtures', 'import-helpers', 'import'),
        reports: [{
                message: `Couldn't find package "tslib".`,
                severity: utils_types_1.Severity.error
            }]
    },
    {
        after: (t) => {
            t.context.sandbox.restore();
        },
        before: (t) => {
            const sandbox = sinon.createSandbox();
            sandbox.stub(utils, 'loadPackage').returns({ exists: true });
            t.context.sandbox = sandbox;
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage: utils.loadPackage });
        },
        name: 'Configuration with "compilerOptions.importHelpers = false" should fail',
        path: path.join(__dirname, 'fixtures', 'import-helpers', 'import-false'),
        reports: [
            {
                message: 'The compiler option "importHelpers" should be enabled to reduce the output size.',
                position: { match: 'false' },
                severity: utils_types_1.Severity.warning
            }
        ]
    },
    {
        after: (t) => {
            t.context.sandbox.restore();
        },
        before: (t) => {
            const sandbox = sinon.createSandbox();
            sandbox.stub(utils, 'loadPackage').returns({ exists: true });
            t.context.sandbox = sandbox;
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage: utils.loadPackage });
        },
        name: 'Configuration with no explicit "compilerOptions.importHelpers" should fail',
        path: path.join(__dirname, 'fixtures', 'import-helpers', 'no-import'),
        reports: [
            {
                message: 'The compiler option "importHelpers" should be enabled to reduce the output size.',
                position: { match: 'compilerOptions' },
                severity: utils_types_1.Severity.warning
            }
        ]
    },
    {
        after: (t) => {
            t.context.sandbox.restore();
        },
        before: (t) => {
            const sandbox = sinon.createSandbox();
            sandbox.stub(utils, 'loadPackage').throws(new Error('Not found'));
            t.context.sandbox = sandbox;
            mock('@hint/utils/dist/src/packages/load-package', { loadPackage: utils.loadPackage });
        },
        name: 'Configuration with no explicit "compilerOptions.importHelpers" and no "tslib" installed should fail',
        path: path.join(__dirname, 'fixtures', 'import-helpers', 'no-import'),
        reports: [
            {
                message: 'The compiler option "importHelpers" should be enabled to reduce the output size.',
                position: { match: 'compilerOptions' },
                severity: utils_types_1.Severity.warning
            },
            {
                message: `Couldn't find package "tslib".`,
                severity: utils_types_1.Severity.error
            }
        ]
    }
];
utils_tests_helpers_1.testLocalHint(hintPath, tests, { parsers: ['typescript-config'], serial: true });
