// @hint/hint-strict-transport-security/dist/tests/_common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const mock = require("mock-require");
const utils_create_server_1 = require("@hint/utils-create-server");
const utilsNetwork = require("@hint/utils-network");
exports.OkayMaxAge = 31536000; // a max-age value larger than the minimum
exports.smallMaxAge = 1; // a max-age value less than the minimum
exports.defaultMinimum = 10886400; // default value of minimum
exports.stsHeader = 'strict-transport-security';
// headers that will pass
exports.maxAgeOnlyHeader = { [exports.stsHeader]: `max-age=${exports.OkayMaxAge}` };
exports.includeSubDomainsHeader = { [exports.stsHeader]: `max-age=${exports.OkayMaxAge}; includeSubDomains` };
exports.preloadHeader = { [exports.stsHeader]: `max-age=${exports.OkayMaxAge}; includeSubDomains; preload` };
exports.mixCaseHeader = { [exports.stsHeader]: `Max-Age=${exports.OkayMaxAge}` };
exports.quotedStringHeader = { [exports.stsHeader]: `max-age="${exports.OkayMaxAge}"; includeSubDomains; preload` };
// headers that will fail
exports.tooShortHeader = { [exports.stsHeader]: `max-age=${exports.smallMaxAge}` };
exports.noMaxAgeHeader = { [exports.stsHeader]: `maxage=${exports.OkayMaxAge}; includeSubDomains; preload` };
exports.multipleMaxAgeHeader = { [exports.stsHeader]: `max-age=${exports.OkayMaxAge}; max-age=${exports.OkayMaxAge + 1}` };
exports.multipleincludeSubDomainsHeader = { [exports.stsHeader]: `includeSubDomains; max-age=${exports.OkayMaxAge}; includeSubDomains` };
exports.wrongDelimiterHeader = { [exports.stsHeader]: `max-age=${exports.OkayMaxAge}, includeSubDomains; preload` };
exports.includeUnitMaxAgeHeader = { [exports.stsHeader]: `max-age=${exports.OkayMaxAge}s; includeSubDomains; preload` };
// api response
exports.notPreloadableError = `www subdomain does not support HTTPS`;
exports.preloaded = { status: 'preloaded' };
exports.unknown = { status: 'unknown' };
exports.noErrors = { errors: [] };
exports.hasErrors = { errors: [{ message: exports.notPreloadableError }] };
// error messages
exports.generateTooShortError = (value) => {
    return `'${exports.stsHeader}' header 'max-age' value should be more than ${value}`;
};
exports.noHeaderError = `'${exports.stsHeader}' header was not specified`;
exports.noMaxAgeError = `'${exports.stsHeader}' header requires 'max-age' directive`;
exports.multipleMaxAgeError = `'${exports.stsHeader}' header contains more than one 'max-age'`;
exports.multipleincludeSubDomainsError = `'${exports.stsHeader}' header contains more than one 'includesubdomains'`;
exports.tooShortErrorDefault = exports.generateTooShortError(exports.defaultMinimum);
exports.DelimiterwrongFormatError = `'${exports.stsHeader}' header has the wrong format: max-age=31536000, includesubdomains`;
exports.UnitwrongFormatError = `'${exports.stsHeader}' header has the wrong format: max-age=31536000s`;
exports.statusServiceError = `Error with getting preload status for https://localhost/.`;
exports.preloadableServiceError = `Error with getting preload eligibility for https://localhost/.`;
exports.problemWithVerificationEndpoint = `Error with getting preload status for https://localhost/. There might be something wrong with the verification endpoint.`;
// override favicon headers so that it doesn't report in chrome
exports.faviconHeaderMaxAgeOnly = {
    '/': { content: utils_create_server_1.generateHTMLPage() },
    '/favicon.ico': { headers: { [exports.stsHeader]: `max-age=${exports.OkayMaxAge + 100}` } }
};
exports.generateHTMLPageData = (content) => {
    return {
        content,
        headers: exports.maxAgeOnlyHeader // the page itself should pass
    };
};
exports.htmlPageWithScriptData = exports.generateHTMLPageData(utils_create_server_1.generateHTMLPage(undefined, '<script src="test.js"></script>'));
exports.htmlPageWithManifestData = exports.generateHTMLPageData(utils_create_server_1.generateHTMLPage('<link rel="manifest" href="test.webmanifest">'));
exports.requestJSONAsyncMock = (responseObject) => {
    const isHTTPS = () => {
        return true;
    };
    const isRegularProtocol = () => {
        return true;
    };
    const normalizeString = (str = '') => {
        return str.toLowerCase();
    };
    const requestJSONAsync = (uri) => {
        let response;
        if (uri.includes('/api/v2/preloadable')) {
            response = responseObject.preloadable;
        }
        else {
            response = responseObject.status;
        }
        if (!response) {
            return Promise.reject('Error with the verification service.');
        }
        return Promise.resolve(response);
    };
    utilsNetwork.isRegularProtocol = isRegularProtocol;
    utilsNetwork.isHTTPS = isHTTPS;
    utilsNetwork.requestJSONAsync = requestJSONAsync;
    mock('@hint/utils-network', utilsNetwork);
    mock('@hint/utils-string', { normalizeString });
};
