// @hint/parser-manifest/dist/src/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("hint/dist/src/lib/types");
const utils_string_1 = require("@hint/utils-string");
const utils_network_1 = require("@hint/utils-network");
const utils_json_1 = require("@hint/utils-json");
// Using `require` instead of `loadJSONFile` so this can be bundled with `extension-browser`.
const schema = require('./schema.json');
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
class ManifestParser extends types_1.Parser {
    constructor(engine) {
        super(engine, 'manifest');
        // Event names.
        this.fetchEndEventName = 'fetch::end::manifest';
        this.fetchErrorEventName = 'fetch::error::manifest';
        this.fetchStartEventName = 'fetch::start::manifest';
        this.parseEndEventName = 'parse::end::manifest';
        this.parseErrorSchemaEventName = 'parse::error::manifest::schema';
        this.parseJSONErrorEventName = 'parse::error::manifest::json';
        // - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
        this.fetchedManifests = new Set();
        engine.on('element::link', this.fetchManifest.bind(this));
        engine.on('fetch::end::manifest', this.validateManifest.bind(this));
        engine.on('scan::end', this.onScanEnd.bind(this));
    }
    onScanEnd() {
        // Clear cached manifests so multiple runs work (e.g. in local connector).
        this.fetchedManifests.clear();
    }
    async fetchManifest(elementFound) {
        const { element, resource } = elementFound;
        /*
         * Check if the target is a local file, and if it is,
         * don't try to fetch the web app manifest file as it's
         * `href` value will most probably not map to anything
         * on disk and the request will fail.
         *
         * TODO: Remove this once things work as expected.
         */
        if (!utils_network_1.isHTTP(resource) && !utils_network_1.isHTTPS(resource)) {
            return;
        }
        // Check if the `link` tag is for the web app manifest.
        if (utils_string_1.normalizeString(element.getAttribute('rel')) !== 'manifest') {
            return;
        }
        // If so, check if a non-empty `href` was specified.
        const hrefValue = utils_string_1.normalizeString(element.getAttribute('href'));
        if (!hrefValue) {
            return;
        }
        // Try to fetch the web app manifest.
        const manifestURL = element.resolveUrl(hrefValue);
        /*
         * Don't fetch the manifest if it has already been fetched (as some
         * connectors fetch automatically).
         *
         * Note: This expects manifest fetches by a connector to occur first.
         * In theory a redundant fetch can still occur if the connector fetch
         * occurs second. This does not happen in practice as this check runs
         * after the page has been loaded (as part of the DOM traversal phase).
         */
        if (this.fetchedManifests.has(manifestURL)) {
            return;
        }
        await this.engine.emitAsync(this.fetchStartEventName, { resource });
        let manifestNetworkData;
        try {
            manifestNetworkData = await this.engine.fetchContent(manifestURL);
        }
        catch (error) {
            await this.engine.emitAsync(this.fetchErrorEventName, {
                element,
                error,
                hops: [manifestURL],
                resource: manifestURL
            });
            return;
        }
        // If the web app manifest was fetched successfully and hasn't already been seen.
        if (this.fetchedManifests.has(manifestURL)) {
            return;
        }
        await this.engine.emitAsync(this.fetchEndEventName, {
            element,
            request: manifestNetworkData.request,
            resource: manifestURL,
            response: manifestNetworkData.response
        });
    }
    async validateManifest(fetchEnd) {
        const { resource, response } = fetchEnd;
        if (this.fetchedManifests.has(resource)) {
            return;
        }
        this.fetchedManifests.add(resource);
        if (response.statusCode >= 400) {
            return;
        }
        await this.engine.emitAsync(`parse::start::manifest`, { resource });
        let result;
        /*
         * Try to see if the content of the web app manifest file
         * is a valid JSON.
         */
        try {
            result = utils_json_1.parseJSON(response.body.content);
        }
        catch (e) {
            await this.engine.emitAsync(this.parseJSONErrorEventName, {
                error: e,
                resource
            });
            return;
        }
        /*
         * Try to see if the content of the web app manifest file
         * is a valid acording to the schema.
         */
        const validationResult = utils_json_1.validate(schema, result.data, result.getLocation);
        if (!validationResult.valid) {
            await this.engine.emitAsync(this.parseErrorSchemaEventName, {
                error: new Error('Invalid manifest'),
                errors: validationResult.errors,
                groupedErrors: validationResult.groupedErrors,
                prettifiedErrors: validationResult.prettifiedErrors,
                resource
            });
            return;
        }
        /*
         * If it is, return the parsed content among with
         * other useful information about the manifest.
         */
        await this.engine.emitAsync(this.parseEndEventName, {
            getLocation: result.getLocation,
            parsedContent: validationResult.data,
            resource
        });
    }
}
exports.default = ManifestParser;
