// @hint/hint-sri/dist/src/meta.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_types_1 = require("@hint/utils-types");
const hint_scope_1 = require("hint/dist/src/lib/enums/hint-scope");
const types_1 = require("./types");
const i18n_import_1 = require("./i18n.import");
const meta = {
    docs: {
        category: utils_types_1.Category.security,
        description: i18n_import_1.getMessage('description', 'en'),
        name: i18n_import_1.getMessage('name', 'en')
    },
    /* istanbul ignore next */
    getDescription(language) {
        return i18n_import_1.getMessage('description', language);
    },
    /* istanbul ignore next */
    getName(language) {
        return i18n_import_1.getMessage('name', language);
    },
    id: 'sri',
    schema: [{
            additionalProperties: false,
            properties: {
                baseline: {
                    enum: Object.keys(types_1.Algorithms).filter((key) => {
                        return isNaN(parseInt(key, 10));
                    }),
                    type: 'string'
                },
                originCriteria: {
                    enum: Object.keys(types_1.OriginCriteria).filter((key) => {
                        return isNaN(parseInt(key, 10));
                    }),
                    type: 'string'
                }
            }
        }],
    scope: hint_scope_1.HintScope.any
};
exports.default = meta;
