// @hint/parser-javascript/dist/src/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const acorn_1 = require("acorn");
const utils_debug_1 = require("@hint/utils-debug");
const content_type_1 = require("@hint/utils/dist/src/content-type");
const types_1 = require("hint/dist/src/lib/types");
const walk_1 = require("./walk");
const debug = utils_debug_1.debug(__filename);
const jsx = require('acorn-jsx');
const jsParser = acorn_1.Parser.extend();
const jsxParser = jsx ? acorn_1.Parser.extend(jsx()) : acorn_1.Parser.extend();
class JavascriptParser extends types_1.Parser {
    constructor(engine) {
        super(engine, 'javascript');
        engine.on('fetch::end::script', this.parseJavascript.bind(this));
        engine.on('fetch::end::unknown', this.onFetchUnknown.bind(this));
        engine.on('element::script', this.parseJavascriptTag.bind(this));
    }
    async emitScript(parser, sourceCode, resource, element) {
        try {
            await this.engine.emitAsync(`parse::start::javascript`, { resource });
            const options = { locations: true, ranges: true };
            const ast = parser.parse(sourceCode, options);
            const tokens = [...parser.tokenizer(sourceCode, options)];
            await walk_1.combineWalk(async (walk) => {
                await this.engine.emitAsync(`parse::end::javascript`, {
                    ast,
                    element,
                    resource,
                    sourceCode,
                    tokens,
                    walk
                });
            });
        }
        catch (err) {
            debug(`Error parsing JS code (${err}): ${sourceCode}`);
        }
    }
    async onFetchUnknown(fetchEnd) {
        if (fetchEnd.response.mediaType !== 'text/jsx') {
            return;
        }
        const code = fetchEnd.response.body.content;
        const resource = fetchEnd.resource;
        await this.emitScript(jsxParser, code, resource, null);
    }
    async parseJavascript(fetchEnd) {
        const code = fetchEnd.response.body.content;
        const resource = fetchEnd.resource;
        await this.emitScript(jsParser, code, resource, null);
    }
    hasSrcAttribute(element) {
        const src = element.getAttribute('src');
        return !!src;
    }
    isJavaScriptType(element) {
        const type = content_type_1.determineMediaTypeForScript(element);
        return !!type;
    }
    async parseJavascriptTag({ element, resource }) {
        if (this.hasSrcAttribute(element)) {
            // Ignore because this will be (or have been) processed in the event 'fetch::end::script'.
            return;
        }
        if (!this.isJavaScriptType(element)) {
            // Ignore if it is not javascript.
            return;
        }
        await this.emitScript(jsParser, element.innerHTML, resource, element);
    }
}
exports.default = JavascriptParser;
