// @hint/parser-babel-config/dist/src/parser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const cloneDeep = require("lodash/cloneDeep");
const hint_1 = require("hint");
const utils_json_1 = require("@hint/utils-json");
const utils_fs_1 = require("@hint/utils-fs");
class BabelConfigParser extends hint_1.Parser {
    constructor(engine) {
        super(engine, 'babel-config');
        this.schema = utils_fs_1.loadJSONFile(path.join(__dirname, 'schema.json'));
        /**
         * package.json => type: 'json' (file type from extention).
         */
        engine.on('fetch::end::json', this.parseBabelConfig.bind(this));
    }
    async validateSchema(config, resource, result) {
        const validationResult = utils_json_1.validate(this.schema, config, result.getLocation);
        const valid = validationResult.valid;
        if (!valid) {
            await this.engine.emitAsync('parse::error::babel-config::schema', {
                error: new Error('Invalid Babel configuration'),
                errors: validationResult.errors,
                groupedErrors: validationResult.groupedErrors,
                prettifiedErrors: validationResult.prettifiedErrors,
                resource
            });
        }
        return validationResult;
    }
    async parseBabelConfig(fetchEnd) {
        const resource = fetchEnd.resource;
        const resourceFileName = path.basename(resource);
        const isPackageJson = resourceFileName === 'package.json';
        const isBabelrc = resourceFileName === '.babelrc';
        if (!isBabelrc && !isPackageJson) {
            return;
        }
        let config;
        try {
            const response = fetchEnd.response;
            // When using local connector to read local files, 'content' is empty.
            let result = utils_json_1.parseJSON(response.body.content, 'extends');
            if (isPackageJson && !result.data.babel) {
                return;
            }
            await this.engine.emitAsync('parse::start::babel-config', { resource });
            // `result.scope('babel')` won't be null since `result.data.babel` was confirmed to exist above.
            result = isPackageJson ? result.scope('babel') : result;
            config = result.data;
            const originalConfig = cloneDeep(config);
            const finalConfig = utils_json_1.finalConfig(config, resource);
            if (finalConfig instanceof Error) {
                await this.engine.emitAsync(`parse::error::babel-config::extends`, {
                    error: finalConfig,
                    getLocation: result.getLocation,
                    resource
                });
                return;
            }
            if (!finalConfig) {
                return;
            }
            config = finalConfig;
            const validationResult = await this.validateSchema(config, resource, result);
            if (!validationResult.valid) {
                return;
            }
            await this.engine.emitAsync('parse::end::babel-config', {
                config: validationResult.data,
                getLocation: result.getLocation,
                originalConfig,
                resource
            });
        }
        catch (err) {
            await this.engine.emitAsync('parse::error::babel-config::json', {
                error: err,
                resource
            });
        }
    }
}
exports.default = BabelConfigParser;
