// @hint/hint-x-content-type-options/dist/tests/tests.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const utils_types_1 = require("@hint/utils-types");
const utils_create_server_1 = require("@hint/utils-create-server");
const utils_tests_helpers_1 = require("@hint/utils-tests-helpers");
// Page data.
const pageWithAlternateStylesheet = utils_create_server_1.generateHTMLPage('<link rel="  alternate stylesheet " href="test.css">');
const pageWithScript = utils_create_server_1.generateHTMLPage(undefined, '<script src="test.js"></script>');
const pageWithScriptAndStylesheet = utils_create_server_1.generateHTMLPage('<link rel="stylesheet" href="test.css">', '<script src="test.js"></script>');
const pageWithStylesheet = utils_create_server_1.generateHTMLPage('<link rel="stylesheet" href="test.css">');
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// Error messages.
const noHeaderErrorMessage = `Response should include 'x-content-type-options' header.`;
const generateInvalidValueMessage = (value = '') => {
    return `'x-content-type-options' header value should be 'nosniff', not '${value}'.`;
};
// - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// Tests.
const tests = [
    {
        name: `HTML page is served without 'X-Content-Type-Options' header`,
        reports: [{
                message: noHeaderErrorMessage,
                severity: utils_types_1.Severity.error
            }],
        serverConfig: {
            '/': { content: utils_create_server_1.generateHTMLPage() },
            '/favicon.ico': { headers: { 'X-Content-Type-Options': 'nosniff' } }
        }
    },
    {
        name: `Favicon is served without 'X-Content-Type-Options' header`,
        reports: [{
                message: noHeaderErrorMessage,
                severity: utils_types_1.Severity.error
            }],
        serverConfig: {
            '/': { content: utils_create_server_1.generateHTMLPage(), headers: { 'Content-Type': 'text/html', 'X-Content-Type-Options': 'nosniff' } },
            '/favicon.ico': ''
        }
    },
    {
        name: `Stylesheet is served without 'X-Content-Type-Options' header`,
        reports: [{
                message: noHeaderErrorMessage,
                severity: utils_types_1.Severity.error
            }],
        serverConfig: {
            '/': { content: pageWithStylesheet, headers: { 'Content-Type': 'text/html', 'X-Content-Type-Options': 'nosniff' } },
            '/favicon.ico': { headers: { 'X-Content-Type-Options': 'nosniff' } },
            '/test.css': ''
        }
    },
    {
        name: `Alternate stylesheet is served without 'X-Content-Type-Options' header`,
        reports: [{
                message: noHeaderErrorMessage,
                severity: utils_types_1.Severity.error
            }],
        serverConfig: {
            '/': { content: pageWithAlternateStylesheet, headers: { 'Content-Type': 'text/html', 'X-Content-Type-Options': 'nosniff' } },
            '/favicon.ico': { headers: { 'X-Content-Type-Options': 'nosniff' } },
            '/test.css': ''
        }
    },
    {
        name: `Resource is specified as a data URI`,
        serverConfig: {
            '/': {
                content: utils_create_server_1.generateHTMLPage(undefined, '<img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAYAAAAfFcSJAAAACklEQVR4nGMAAQAABQABDQottAAAAABJRU5ErkJggg==">'),
                headers: { 'X-Content-Type-Options': 'nosniff' }
            },
            '/favicon.ico': { headers: { 'X-Content-Type-Options': 'nosniff' } }
        }
    },
    {
        name: `Script is served with 'X-Content-Type-Options' header with invalid value`,
        reports: [{
                message: generateInvalidValueMessage('invalid'),
                severity: utils_types_1.Severity.error
            }],
        serverConfig: {
            '/': { content: pageWithScript, headers: { 'Content-Type': 'text/html', 'X-Content-Type-Options': 'nosniff' } },
            '/favicon.ico': { headers: { 'X-Content-Type-Options': 'nosniff' } },
            '/test.js': { headers: { 'X-Content-Type-Options': 'invalid' } }
        }
    },
    {
        name: `All resources are served with 'X-Content-Type-Options' header`,
        serverConfig: { '*': { content: pageWithScriptAndStylesheet, headers: { 'Content-Type': 'text/html', 'X-Content-Type-Options': 'nosniff' } } }
    }
];
utils_tests_helpers_1.testHint(utils_tests_helpers_1.getHintPath(__filename), tests);
