// @hint/utils/dist/src/npm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const child_process_1 = require("child_process");
const path = require("path");
const fs = require("fs");
const npmRegistryFetch = require("npm-registry-fetch");
const utils_debug_1 = require("@hint/utils-debug");
const utils_fs_1 = require("@hint/utils-fs");
const logger = require("./logging");
const packages_1 = require("./packages");
const has_yarnlock_1 = require("./has-yarnlock");
const debug = utils_debug_1.debug(__filename);
const install = (command) => {
    return new Promise((resolve, reject) => {
        const npmInstall = child_process_1.spawn(command, [], { shell: true, stdio: 'inherit' });
        npmInstall.on('error', (err) => {
            return reject(err);
        });
        npmInstall.on('exit', (code) => {
            if (code !== 0) {
                return reject();
            }
            return resolve(true);
        });
    });
};
/** Install the given packages. */
exports.installPackages = async (packages) => {
    /** Whether or not the package should be installed as devDependencies. */
    let isDev = false;
    /** Current working directory. */
    const currentWorkingDir = utils_fs_1.cwd();
    /** Whether or not the process is running in windows */
    const isWindows = process.platform === 'win32';
    if (packages.length === 0) {
        return Promise.resolve(true);
    }
    const hintLocalPath = path.join(currentWorkingDir, 'node_modules', 'hint', 'package.json');
    // Check if hint is installed locally.
    const global = !fs.existsSync(hintLocalPath); // eslint-disable-line no-sync
    /** package manager to install the packages. */
    const packageManagerChoice = (!global && await has_yarnlock_1.hasYarnLock(currentWorkingDir)) ? 'yarn' : 'npm';
    if (!global) {
        try {
            const packagePath = packages_1.findPackageRoot(currentWorkingDir);
            const jsonContent = utils_fs_1.loadJSONFile(path.join(packagePath, 'package.json'));
            // If `hint` is a devDependency, then set all packages as devDependencies.
            isDev = jsonContent.devDependencies && jsonContent.devDependencies.hasOwnProperty('hint');
        }
        catch (err) {
            // Even if `hint` is installed locally, package.json could not exist in the current working directory.
            isDev = false;
        }
    }
    const installCommand = {
        npm: `npm install${global ? ' --global' : ''}${isDev ? ' --save-dev' : ''}`,
        yarn: `yarn add${isDev ? ' --dev' : ''}`
    };
    const command = `${installCommand[packageManagerChoice]} ${packages.join(' ')}`;
    try {
        debug(`Running command ${command}`);
        logger.log('Installing packages...');
        await install(command);
        return true;
    }
    catch (err) {
        debug(err);
        /*
         * There was an error installing packages.
         * Show message to install packages manually (maybe permissions error?).
         */
        /* istanbul ignore next */
        logger.error(`
There was a problem installing packages.
Please try executing:
    ${!isWindows && global ? 'sudo ' : ''}${command}
            manually to install all the packages.`);
        return false;
    }
};
/** Filters the packages that `startsWith` `initTerm`. */
const filterPackages = (packages, initTerm) => {
    return packages.filter((pkg) => {
        return pkg.name.startsWith(initTerm);
    });
};
/** Get npm packages from the object returned for npmRegistryFetch.json. */
const getPackages = (result) => {
    return result.objects.map((obj) => {
        return obj.package;
    });
};
/** Generate a search query to search packages. */
const generateSearchQuery = (searchTerm, from, size = 100) => {
    return `/-/v1/search?text=${searchTerm}&size=${size}${from ? `&from=${from}` : ''}`;
};
/**
 * Searches all the packages in npm given `searchTerm`.
 */
exports.search = async (searchTerm) => {
    const result = (await npmRegistryFetch.json(generateSearchQuery(searchTerm)));
    let total = getPackages(result);
    while (result.total > total.length) {
        const r = (await npmRegistryFetch.json(generateSearchQuery(searchTerm, total.length)));
        total = total.concat(getPackages(r));
    }
    return total;
};
/** Get core packages from npm. */
exports.getOfficialPackages = async (type) => {
    const hints = await exports.search(`@hint/${type}`);
    /*
     * We need to filter the results because the search can
     * include other packages that doesn't start with `@hint/{type}`.
     */
    return filterPackages(hints, `@hint/${type}`);
};
/** Get external packages from npm. */
exports.getUnofficialPackages = async (type) => {
    const hints = await exports.search(`webhint-${type}`);
    /*
     * We need to filter the results because the search can
     * include other packages that doesn't start with `webhint-{type}`.
     */
    return filterPackages(hints, `webhint-${type}`);
};
