// @hint/utils-json/dist/src/parse-json.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const jsonc_parser_1 = require("jsonc-parser");
const rxIsNumber = /^[0-9]+$/;
class JSONResult {
    constructor(data, root, lines, alternatePath) {
        this._data = data;
        this._lines = lines;
        this._root = root;
        this._alternatePath = alternatePath;
        // Ensure `getLocation` can be passed around without losing context
        this.getLocation = this.getLocation.bind(this);
    }
    get data() {
        return this._data;
    }
    getLocation(path, options) {
        const segments = this.pathToSegments(path);
        let node = null;
        while (!node && segments.length > 0) {
            /* istanbul ignore next */
            node = jsonc_parser_1.findNodeAtLocation(this._root, segments) || null;
            segments.pop();
        }
        /**
         * The node isn't in the current file. Use alternative path if provided. This happens
         * when extending configurations.
         */
        /* istanbul ignore next */
        if (!node && this._alternatePath && path !== this._alternatePath) {
            return this.getLocation(this._alternatePath, options);
        }
        return this.offsetToLocation(this.getAdjustedOffset(node, path, options));
    }
    /* istanbul ignore next */
    scope(path) {
        const segments = this.pathToSegments(path);
        const node = jsonc_parser_1.findNodeAtLocation(this._root, segments);
        const value = this.findValueAtLocation(segments);
        return node ? new JSONResult(value, node, this._lines) : null;
    }
    /**
     * Determine the best offset to point to given the provided context.
     */
    getAdjustedOffset(node, path, options) {
        // Point to the root if nothing better is available
        /* istanbul ignore if */
        if (!node) {
            return this._root.offset;
        }
        // Point to the value if requested (default location returned by jsonc-parser)
        /* istanbul ignore next */
        if (options && options.at === 'value') {
            return node.offset;
        }
        // Point to the value for array items
        /* istanbul ignore if */
        if (path.endsWith(']')) {
            return node.offset;
        }
        // Point to the value if there's no parent
        /* istanbul ignore if */
        if (!node.parent) {
            return node.offset;
        }
        // Otherwise point to the name
        return node.parent.offset + 1; // +1 to get past the quote (")
    }
    /**
     * Find the value at the given path in the JSON DOM.
     * @param segments The path to the value.
     */
    /* istanbul ignore next */
    findValueAtLocation(segments) {
        let value = this._data;
        segments.forEach((segment) => {
            value = value[segment];
        });
        return value;
    }
    /**
     * Convert a source offset into a `ProblemLocation` with line/column data.
     * @param offset The offset in the original source.
     */
    offsetToLocation(offset) {
        for (let i = 0, n = 0; i < this._lines.length; i++) {
            const lineLength = this._lines[i].length;
            if (offset <= n + lineLength) {
                return {
                    column: offset - n,
                    line: i
                };
            }
            // Move to the next line (+1 to account for the newline)
            n += lineLength + 1;
        }
        return null;
    }
    /**
     * Convert a JS-style path string to the `Segment` array needed by jsonc-parser.
     * @param path The path to convert (e.g. `foo.items[1].bar`).
     * @returns An array of `Segment` properties (e.g. `['foo', 'items', 1, 'bar']`).
     */
    pathToSegments(path) {
        return path
            // Strip leading dot (.) if present (ajv bug?)
            .replace(/^\./, '')
            // Ignore trailing `]` from `foo[1]`
            .replace(/]/g, '')
            // Break items on `.` or `[`
            .split(/[[.]/)
            // Ensure numbers are not returned as strings
            .map((k) => {
            /* istanbul ignore next */
            return rxIsNumber.test(k) ? parseInt(k) : k;
        });
    }
}
/**
 * Parse the provided JSON returning a `JSONResult` with location information.
 * @param json The JSON string to parse
 * @param alternatePath The alternative `path` to use when one is not found
 */
exports.parseJSON = (json, alternatePath) => {
    const lines = json.split('\n');
    const data = jsonc_parser_1.parse(json);
    const root = jsonc_parser_1.parseTree(json);
    // If we didn't get a root, it's invalid JSON
    /* istanbul ignore if */
    if (!root) {
        // Use the built-in JSON parser to get an error
        JSON.parse(json);
    }
    return new JSONResult(data, root, lines, alternatePath);
};
