// @hint/utils-json/dist/src/schema-validator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ajv = require("ajv");
const cloneDeep = require("lodash/cloneDeep");
const forEach = require("lodash/forEach");
const groupBy = require("lodash/groupBy");
const reduce = require("lodash/reduce");
const without = require("lodash/without");
/*
 * If we want to use the ajv types in TypeScript, we need to import
 * ajv in a lowsercase variable 'ajv', otherwise, we can't use types
 * like `ajv.Ajv'.
 */
const validator = new ajv({
    $data: true,
    allErrors: true,
    logger: false,
    useDefaults: true,
    verbose: true
});
var ErrorKeyword;
(function (ErrorKeyword) {
    ErrorKeyword["additionalProperties"] = "additionalProperties";
    ErrorKeyword["anyOf"] = "anyOf";
    ErrorKeyword["enum"] = "enum";
    ErrorKeyword["oneOf"] = "oneOf";
    ErrorKeyword["pattern"] = "pattern";
    ErrorKeyword["required"] = "required";
    ErrorKeyword["type"] = "type";
    ErrorKeyword["uniqueItems"] = "uniqueItems";
})(ErrorKeyword || (ErrorKeyword = {}));
const generateError = (type, action) => {
    return (error, errors) => {
        if (error.keyword !== type) {
            return null;
        }
        const property = error.dataPath.substr(1);
        return action(error, property, errors);
    };
};
/**
 * Returns a readable error for 'required' errors.
 */
const generateRequiredError = generateError(ErrorKeyword.required, (error, property) => {
    return `'${property ? property : 'root'}' ${error.message}`;
});
/**
 * Returns a readable error for 'additionalProperty' errors.
 */
const generateAdditionalPropertiesError = generateError(ErrorKeyword.additionalProperties, (error, property) => {
    const additionalProperty = error.params.additionalProperty;
    return `'${property ? property : 'root'}' ${property ? error.message : `${error.message}`}. Additional property found '${additionalProperty}'.`;
});
/**
 * Returns a readable error for 'enum' errors.
 */
const generateEnumError = generateError(ErrorKeyword.enum, (error, property) => {
    const allowedValues = error.params.allowedValues;
    return `'${property}' ${error.message} '${allowedValues.join(', ')}'. Value found '${error.data}'`;
});
/**
 * Returns a readable error for 'pattern' errors.
 */
const generatePatternError = generateError(ErrorKeyword.pattern, (error, property) => {
    return `'${property}' ${error.message && error.message.replace(/"/g, '\'')}. Value found '${error.data}'`;
});
/**
 * Returns a readable error for 'type' errors.
 */
const generateTypeError = generateError(ErrorKeyword.type, (error, property) => {
    return `'${property}' should be '${error.params.type}'.`;
});
const generateAnyOfError = generateError(ErrorKeyword.anyOf, (error, property, errors) => {
    const otherErrors = without(errors, error);
    const results = otherErrors.map((otherError) => {
        // eslint-disable-next-line @typescript-eslint/no-use-before-define, no-use-before-define
        return generate(otherError);
    });
    return results.join(' or ');
});
const generateUniqueItemError = generateError(ErrorKeyword.uniqueItems, (error, property) => {
    return `'${property}' ${error.message && error.message.replace(/"/g, '\'')}.`;
});
const getRequiredProperty = (error) => {
    return `'${error.params.missingProperty}'`;
};
const getTypeProperty = (error) => {
    return `'${error.params.type}'`;
};
const getEnumValues = (error) => {
    return `'${error.params.allowedValues.join(', ')}'`;
};
const generateAnyOfMessageRequired = (errors) => {
    return `should have required ${errors.length === 1 ? 'property' : 'properties'} ${errors.map(getRequiredProperty).join(' or ')}`;
};
const generateAnyOfMessageType = (errors) => {
    return `should be ${errors.map(getTypeProperty).join(' or ')}.`;
};
const generateAnyOfMessageEnum = (errors) => {
    return `should be equal to one of the allowed values ${errors.map(getEnumValues).join(' or ')}. Value found '${JSON.stringify(errors[0].data)}'.`;
};
const generateAnyOfMessage = {
    [ErrorKeyword.required]: generateAnyOfMessageRequired,
    [ErrorKeyword.type]: generateAnyOfMessageType,
    [ErrorKeyword.enum]: generateAnyOfMessageEnum
};
const errorGenerators = [generateAdditionalPropertiesError, generateEnumError, generatePatternError, generateTypeError, generateUniqueItemError, generateRequiredError, generateAnyOfError];
/**
 * Returns a readable error message.
 */
const generate = (error, errors) => {
    return errorGenerators.reduce((message, generator) => {
        const newErrorMessage = generator(error, errors);
        if (newErrorMessage) {
            return newErrorMessage;
        }
        return message;
    }, error.message || '');
};
/**
 * Returns a readable error for 'anyOf' and 'oneOf' errors.
 * e.g.: 'root' should have required properties 'connector' or 'extends'
 */
const generateAnyOfGroupedError = (error, errors) => {
    const otherErrors = without(errors, error);
    const grouped = groupBy(otherErrors, 'keyword');
    const results = reduce(grouped, (allMessages, groupedErrors, keyword) => {
        const dataPath = error.dataPath;
        const messageGenerator = generateAnyOfMessage[keyword];
        if (messageGenerator) {
            allMessages.push(`'${dataPath ? dataPath.substr(1) : 'root'}' ${messageGenerator(groupedErrors)}`);
            return allMessages;
        }
        groupedErrors.forEach((error) => {
            /* istanbul ignore next */
            const errorGenerated = generate(error, groupedErrors) || '';
            /* istanbul ignore else */
            if (errorGenerated) {
                allMessages.push(`${errorGenerated}`);
            }
        });
        return allMessages;
    }, []);
    return results.join(' Or ');
};
const generateErrorsMessage = (errors) => {
    const grouped = groupBy(errors, 'keyword');
    const result = reduce(grouped, (allMessages, groupedErrors, keyword) => {
        if (keyword === ErrorKeyword.required) {
            const dataPath = groupedErrors[0].dataPath;
            allMessages.push(`'${dataPath ? dataPath.substr(1) : 'root'}' should have required ${groupedErrors.length === 1 ? 'property' : 'properties'} ${groupedErrors.map(getRequiredProperty).join(' and ')}`);
            return allMessages;
        }
        groupedErrors.forEach((error) => {
            allMessages.push(generate(error, groupedErrors) || '');
        });
        return allMessages;
    }, []);
    return result;
};
/**
 * Group messages with the same data path.
 * e.g.:
 * * Input (only messages):
 * *   - should be equal to one of the allowed values (dataPath: ".hints['amp-validator']")
 * *   - should be number (dataPath: ".hints['amp-validator'])"
 * *   - should be equal to one of the allowed values (dataPath: ".hints['amp-validator']")
 * *   - should have required property 'connector' (dataPath: "")
 * *   - should have required property 'extends' (dataPath: "")
 * *   - should match some schema in anyOf (dataPath: "")
 *
 * * Output (only messages):
 * *   - 'hints['amp-validator']' should be equal to one of the allowed values 'off, warning, error' or '0, 1, 2'. Value found '"notvalid"'. Or 'hints['amp-validator']' should be 'number'.
 * *   - 'root' should have required properties 'connector' or 'extends'
 */
const groupMessages = (errors) => {
    const grouped = groupBy(errors, 'dataPath');
    const result = reduce(grouped, (allErrors, groupErrors) => {
        let errors = groupErrors;
        const anyOf = groupErrors.find((error) => {
            return error.keyword === ErrorKeyword.anyOf || error.keyword === ErrorKeyword.oneOf;
        });
        if (anyOf) {
            const anyOfErrors = groupErrors.filter((error) => {
                /* istanbul ignore next */
                return error.schemaPath.includes(anyOf.schemaPath) || anyOf.schema.some((schema) => {
                    return error.schemaPath.includes(schema.$ref);
                });
            });
            errors = without(groupErrors, ...anyOfErrors);
            allErrors.push({
                errors: anyOfErrors,
                location: anyOfErrors[0].location,
                message: generateAnyOfGroupedError(anyOf, anyOfErrors)
            });
            /* istanbul ignore else */
            if (errors.length === 0) {
                return allErrors;
            }
        }
        /*
         * If there is no 'anyOf' error, then join with 'and' the rest of the messages.
         * if they have the same keyword.
         */
        const groupedByLocation = groupBy(errors, (error) => {
            /* istanbul ignore if */
            if (error.location) {
                return `column${error.location.column}row${error.location.column}`;
            }
            return '-';
        });
        forEach(groupedByLocation, (group) => {
            allErrors.push({
                errors: group,
                location: group[0].location,
                message: generateErrorsMessage(group).join(' and ')
            });
        });
        return allErrors;
        // return allMessages;
    }, []);
    return result;
};
/**
 * Add location information to the provided schema error object.
 */
/* istanbul ignore next */
const errorWithLocation = (error, getLocation) => {
    let path = error.dataPath;
    const additionalProperty = error.params && error.params.additionalProperty;
    if (additionalProperty) {
        path = path ? `${path}.${additionalProperty}` : additionalProperty;
    }
    return Object.assign(error, { location: getLocation(path.replace(/'/g, '')) || undefined });
};
const prettify = (errors) => {
    const grouped = groupBy(errors, 'dataPath');
    const result = reduce(grouped, (allMessages, groupErrors) => {
        groupErrors.forEach((error) => {
            allMessages.push(generate(error, groupErrors) || '');
        });
        return allMessages;
    }, []);
    return result;
};
exports.validate = (schema, json, getLocation) => {
    // We clone the incoming data because the validator can modify it.
    const data = cloneDeep(json);
    const validateFunction = validator.compile(schema);
    const valid = validateFunction(data);
    let errors = validateFunction.errors || [];
    /* istanbul ignore if */
    if (errors && getLocation) {
        errors = errors.map((e) => {
            return errorWithLocation(e, getLocation);
        });
    }
    const prettifiedErrors = prettify(errors);
    const groupedErrors = groupMessages(errors);
    return {
        data,
        errors,
        groupedErrors,
        prettifiedErrors,
        valid
    };
};
