// @hint/utils-dom/dist/src/htmldocument.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const url_1 = require("url");
const parse5 = require("parse5");
const htmlparser2Adapter = require("parse5-htmlparser2-tree-adapter");
const cssSelect = require("css-select");
const htmlelement_1 = require("./htmlelement");
const CACHED_CSS_SELECTORS = new Map();
class HTMLDocument {
    constructor(document, finalHref, originalDocument) {
        this._pageHTML = '';
        this._document = document;
        this._documentElement = this.findDocumentElement();
        this.originalDocument = originalDocument;
        this._pageHTML = parse5.serialize(document, { treeAdapter: htmlparser2Adapter });
        this._base = this.getBaseUrl(finalHref);
    }
    findDocumentElement() {
        return this._document.children.find((node) => {
            return node.type === 'tag' && node.name === 'html';
        });
    }
    getBaseUrl(finalHref) {
        const baseElement = this.querySelectorAll('base[href]')[0];
        const baseHref = baseElement ? baseElement.getAttribute('href') : null;
        if (!baseHref) {
            return new url_1.URL(finalHref).href;
        }
        return new url_1.URL(baseHref, finalHref).href;
    }
    get documentElement() {
        return new htmlelement_1.HTMLElement(this._documentElement, this);
    }
    get base() {
        return this._base;
    }
    get compatMode() {
        return this._document['x-mode'] === 'quirks' ?
            'BackCompat' :
            'CSS1Compat';
    }
    get isFragment() {
        return !this.originalDocument && !this._documentElement.sourceCodeLocation;
    }
    pageHTML() {
        return this._pageHTML;
    }
    querySelectorAll(selector) {
        if (!CACHED_CSS_SELECTORS.has(selector)) {
            CACHED_CSS_SELECTORS.set(selector, cssSelect.compile(selector));
        }
        const matches = cssSelect(CACHED_CSS_SELECTORS.get(selector), this._document.children);
        const result = matches.map((element) => {
            return new htmlelement_1.HTMLElement(element, this);
        });
        return result;
    }
    resolveUrl(url) {
        return new url_1.URL(url, this._base).href;
    }
}
exports.HTMLDocument = HTMLDocument;
