// @hint/hint-ssllabs/dist/tests/tests.js
"use strict";
/* eslint sort-keys: 0 */
Object.defineProperty(exports, "__esModule", { value: true });
const mock = require("mock-require");
const utils_tests_helpers_1 = require("@hint/utils-tests-helpers");
const utils_types_1 = require("@hint/utils-types");
const hintPath = utils_tests_helpers_1.getHintPath(__filename);
const ssllabsMock = (response) => {
    const mockedModule = {
        // Original node-ssllabs uses callback and we promisify in the hint
        scan: (options, callback) => {
            if (response === null) {
                return callback('Error');
            }
            return callback(null, response);
        }
    };
    mock('node-ssllabs', mockedModule);
};
const results = {
    aplussite: { endpoints: [{ grade: 'A+' }] },
    asite: {
        endpoints: [{ grade: 'A' }, {
                grade: 'A',
                serverName: 'a-site.net'
            }]
    },
    nohttps: { endpoints: [{ details: { protocols: [] } }] }
};
const testsForDefaults = [
    {
        name: `Site with with A+ grade passes`,
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock(results.aplussite);
        }
    },
    {
        name: `Site A grade passes`,
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock(results.asite);
        }
    },
    {
        name: `Domain without HTTPS fails with default configuration`,
        reports: [{
                message: `'http://example.com/' does not support HTTPS.`,
                severity: utils_types_1.Severity.error
            }],
        serverUrl: 'http://example.com',
        before() {
            ssllabsMock(results.nohttps);
        }
    },
    {
        name: `Resource is not an HTML document`,
        serverConfig: { '/': { headers: { 'Content-Type': 'image/png' } } }
    }
];
const testsForConfigs = [
    {
        name: `Site with A+ grade passes with A+ minimum`,
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock(results.aplussite);
        }
    },
    {
        name: `Site with A grade doesn't pass with A+ minimum`,
        reports: [
            {
                message: `https://example.com/'s grade A does not meet the minimum A+ required.`,
                severity: utils_types_1.Severity.error
            },
            {
                message: `a-site.net's grade A does not meet the minimum A+ required.`,
                severity: utils_types_1.Severity.error
            }
        ],
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock(results.asite);
        }
    },
    {
        name: `Domain without HTTPS fails with a custom configuration`,
        reports: [{
                message: `'http://example.com/' does not support HTTPS.`,
                severity: utils_types_1.Severity.error
            }],
        serverUrl: 'http://example.com',
        before() {
            ssllabsMock(results.nohttps);
        }
    }
];
const testsForErrors = [
    {
        name: 'Issue gettings results from SSL Labs reports error',
        reports: [{
                message: `Could not get results from SSL Labs for 'https://example.com/'.`,
                severity: utils_types_1.Severity.warning
            }],
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock(null);
        }
    },
    {
        name: 'Missing endpoints reports an error',
        reports: [{
                message: `Didn't get any result for https://example.com/.
There might be something wrong with SSL Labs servers.`,
                severity: utils_types_1.Severity.warning
            }],
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock({});
        }
    },
    {
        name: 'Empty endpoints array reports an error',
        reports: [{
                message: `Didn't get any result for https://example.com/.
There might be something wrong with SSL Labs servers.`,
                severity: utils_types_1.Severity.warning
            }],
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock({ endpoints: [] });
        }
    },
    {
        name: 'Response with right status code but nothing inside reports an error',
        reports: [{
                message: `Didn't get any result for https://example.com/.
There might be something wrong with SSL Labs servers.`,
                severity: utils_types_1.Severity.warning
            }],
        serverUrl: 'https://example.com',
        before() {
            ssllabsMock(undefined);
        }
    }
];
utils_tests_helpers_1.testHint(hintPath, testsForDefaults, { serial: true });
utils_tests_helpers_1.testHint(hintPath, testsForConfigs, {
    hintOptions: { grade: 'A+' },
    serial: true
});
utils_tests_helpers_1.testHint(hintPath, testsForErrors, { serial: true });
